lib/tasks_private/lib/aws_instance_data_parser.rb

```diff
@@ -0,0 +1,316 @@
+  end
+
+  ## memory
+
+  def memory
+    @memory ||= begin
+      memory = product_data['memory']
+      (MEMORY_REGEXP.match(memory)&.captures&.first || save_unknown(memory)).tr(',', '').to_f
+    end
+  end
+
+  ## storage
+
+  def storage
+    product_data['storage']
+  end
+
+  def ebs_only?
+    storage == 'EBS only'
+  end
+
+  def ebs_optimized?
+    product_data['ebsOptimized'] == 'Yes' ||
+      product_data['dedicatedEbsThroughput'].present? ||
+      EBS_OPTIMIZED_TYPES.include?(base_type)
+  end
+
+  def storage_volumes
+    parsed_storage.volumes
+  end
+
+  def storage_size
+    parsed_storage.size
+  end
+
+  def storage_type
+    parsed_storage.type
+  end
+
+  ## network
+
+  def network_performance
+    net_perf = product_data['networkPerformance']
+    net_perf =~ NETWORK_REGEXP ? :very_high : net_perf.downcase.gsub(/\s/, '_').to_sym
+  end
+
+  def enhanced_networking?
+    product_data['enhancedNetworkingSupported'] == 'Yes' ||
+      ENHANCED_NETWORKING_TYPES.include?(base_type)
+  end
+
+  def clusterable_networking?
+    CLUSTERABLE_TYPES.include?(base_type)
+  end
+
+  def vpc_only?
+    VPC_ONLY_TYPES.include?(base_type)
+  end
+
+  private
+
+  def parse!
+    @instance_data = {
+      :current_generation      => current_generation?,
+      :current_version         => current_version?,
+      :name                    => instance_type,
+      :family                  => instance_family,
+      :description             => description,
+      :memory                  => memory.gigabytes.to_i,
+      :memory_gb               => memory,
+      :vcpu                    => vcpus,
+      :ebs_only                => ebs_only?,
+      :instance_store_size     => storage_size.gigabyte.to_i,
+      :instance_store_size_gb  => storage_size,
+      :instance_store_volumes  => storage_volumes,
+      :instance_store_type     => storage_type,
+      :architecture            => cpu_arches,
+      :virtualization_type     => virtualization_type,
+      :network_performance     => network_performance,
+      :physical_processor      => physical_processor,
+      :processor_clock_speed   => cpu_clock_speed,
+      :intel_aes_ni            => intel_aes_ni?           || nil,
+      :intel_avx               => intel_avx?              || nil,
+      :intel_avx2              => intel_avx2?             || nil,
+      :intel_turbo             => intel_turbo?            || nil,
+      :ebs_optimized_available => ebs_optimized?          || nil,
+      :enhanced_networking     => enhanced_networking?    || nil,
+      :cluster_networking      => clusterable_networking? || nil,
+      :vpc_only                => vpc_only?,
+    }.freeze
+    @unknown_values.freeze
+    @parsed = true
+  end
+
+  def save_unknown(value, attribute_name = nil, nils: nil)
+    attribute_name ||= caller_locations(1..1).first.label # use a bit of magic
+    (@unknown_values[attribute_name.to_sym] ||= Set.new) << value
+    nils ? Array.new(nils, nil) : nil
```

Comment:
I don't really understand the purpose of this method.  I think its purpose is to deal with post-parse-presentation of unexpected things, and then returning an Array of 3 nils (not sure shy that's needed).

Reply:
`save_unknown` used in after `||` part of assignments. e.g. `something = we_have || we_dont_have_and_we_need_to_remember_that`, so I return `nil` to set the variable(s)

Reply:
Oh I see, so it's for learning about new things.  Hmmm... that's interesting.